tests/test_utils.py

```diff
@@ -824,3 +826,124 @@ def test_non_operations_in_list(self, arg):
             ValueError, match="The given operation_list does not only contain Operations"
         ):
             pu.inv(arg)
+
+
+class TestExecutionCounter:
+    """Tests the ExecutionCounter."""
+
+    @pytest.mark.parametrize("dev, expected", [
+        (qml.device('default.qubit', wires=1), 6),
+        (qml.device('default.qubit.tf', wires=1), 6),  # should this device not have the same num of evals as autograd?
```

Comment:
This was puzzling me, and I need to understand it before merging...Would you know why the tf device has a different (and unexpected) number of calls than the autograd device?

I thought those devices do not need to use param shift rules, hence only one call to compute a gradient (as in the autograd one).

Reply:
So even weirder, if I use the tape in a different way, I get the same number (5) for all three devices (assuming `jacobian` does the same thing as `grad`, right? it seems to have 2 executions for each). 

```
params=[0.3]

with qml.tape.JacobianTape() as tape:
    qml.RX(params[0], wires=0)
    qml.Hadamard(wires=0)
    qml.expval(qml.PauliZ(wires=0))

dev = qml.device('default.qubit', wires=[0], analytic=True)
#dev = qml.device('default.qubit.autograd', wires=[0], analytic=True)
#dev = qml.device('default.qubit.tf', wires=[0], analytic=True)

# Tries to match the sequence of calls in the test
with ExecutionCounter() as counter:
    tape.execute(dev, params=[0.1])
    tape.execute(dev, params=[0.5])
    tape.jacobian(dev)
    tape.execute(dev)
    
print(f"counts = {counter.counts}")
```

Reply:
I think this is explainable: if you don't use qnodes at all, the diff method is not picked and therefore set to default, which is finite diff. Finite diff needs 2 evaluations: the original tape and the shifted tape (and not three, like parameter-shift: original tape + 2 shifted tapes).